OpenMage/magento-lts
# Add OpenMage Contributors Copyright

### Description (*)

From discussion #2276

Added `@copyright` header to 4,687 files that have been modified by the OpenMage project since 2015. The OpenMage project has done great work and deserves its copyright notice.

Instead of a generic 2015-2022 year range, each file has its own specific year, or year range, based on the git history of that file. I've included the script that I used to generate this commit.

The script has comments explaining what it does, but here's an overview:

1. Read all the entire commit history of the repo using `git log --pretty='%H %cs %s'`
2. Loop through each commit and try and determine if this was upstream changes from Magento, LLC or if it was a commit from OpenMage contributors. To do this, the script looks for commit messages such as:
	* "Import Magento"
	* "Merge ... magento-1.*"
	* "SUPEE"
	* Also, some hardcoded commit hashes that don't fit a pattern
3. Loop through each file in the entire repo that contains `@copyright` and find the first and last commit hash from our results in step 2
4. Add or update the `@copyright The OpenMage Contributors` line in each file

To see a list of commits attributed to either Magento or OpenMage, run `php shell/update-copyright.php dump`. I will post the results of that in the next comment so you can see the output here, too.

The rationale behind not using a generic date range is for ease of `git diff` between versions as suggested by @fballiano. Instead, if the script is run before a version is tagged, then it will only be modifying files that were already changed by a commit in that version.

Here is the output from when I ran the script:

```console
$ time php shell/update-copyright.php 
Updated 4687 file(s)

real    3m13.286s
user    2m31.269s
sys     0m48.590s
```

It did take a few minutes, maybe the reading of `git log` on each of the 4,000+ files can be optimized.

Also, this PR is made to 1.9.4.x. I can also make a smaller PR to 20.0 since there will be discrepencies. This could create some merge conflicts in the future, but I don't think it will be too bad. For example assume some file has '2015-2019' in 1.9.4.x and '2015-2020' in 20.0 (because there was a 20.0 branch only commit in 2020 to that file). If the file was then changed in 1.9.4.x in 2022, it would conflict I think when v19 is merged into v20. I ask the maintainers to weigh in on that.

### Related Pull Requests


### Fixed Issues (if relevant)


### Manual testing scenarios (*)

Check out this branch and run:

```console
git diff -U0 HEAD~1 | grep '^[+-]' | grep -Ev '^(--- a/|\+\+\+ b/)' | sort -u
```

That will print all unique changes made in the second commit of this PR. You should only see copyright changes. I will post the output in a follow up comment.

### Questions or comments

I have intentionally made two commits in this PR. I think it is best to not squash them.

Other changes from #2276 will be made in a second PR.

### Contribution checklist (*)
